dcos/dcos-ui
# DCOS-48316: Jobs form

Implements a basic new Jobs form. The modal and form are based on the Services implementations (`CreateServiceModal`, `CreateServiceModalForm`). 

Included in these commits:
- Basic tabbed job form (that can toggle to JSON editor) and modal
- General form section component
- Modifications to `MetronomeClient` to upgrade to metronome API v1 spec for job create and update
- Job editing

Not included in this PR:
- The form doesn't open to it's own dedicated route (do we want this? @AmrAbdelrazik)

This PR overlaps a couple different JIRA tickets. These can be separated into separate PRs if necessary but it's difficult to test the changes end to end without all 3:

### DCOS-48316: Jobs form modal
- Adds a modal and form component
- Create/edit buttons open a full page modal
- Form contains one tab (General)
- Basic validation and error visibility is set up
- Modal can toggle a JSON editor

### DCOS-48445: Datalayer changes
- Updates the create and update job mutations to be compliant with the metronome v1 API spec (previously was v0)
- See https://github.com/dcos/dcos-ui/pull/3599/commits/6ac68557ae19eb7b71c9aedbf4c485ec22d75ddb, https://github.com/dcos/dcos-ui/pull/3599/commits/30a00897fb573ae6479e1f5292f2f16116627751, https://github.com/dcos/dcos-ui/pull/3599/commits/5a05f2c8c63dc7e7caf751c28270207a1b1dae46

### DCOS-48352: General form section
- Adds the general form section to enter the bare minimum for a job (ID, description, cpus, mem, disk, gpus, command, container Image)
- Should be able to edit from both form inputs and JSON
- Validation should prevent submission of an invalid form and show errors

## Testing

Different scenarios to try:
- Create a job (it should work)
- When 'Command Only' is chosen, only Command input should be shown below (required)
- When 'Container Image' is chosen, both Container Image (required) and Command inputs should be shown below (also, a `ucr` property should be added to the job JSON - visible in JSON editor)
- Try putting the validation to the test (try submitting an invalid form)
- Try editing using the JSON editor
- Try creating a job and adding properties to the JSON that aren't yet available in the form UI. These should persist when the job is added. For example, try adding a schedule like:
```
{
"job": {...},
"schedule": {
    "id": "my-schedule-data",
    "cron": "0 4 * * *"
  }
}
```


## Trade-offs

This was put together using existing patterns from the Services form. The tradeoff here is using what is tried & tested & available (fast but comes with its own challenges and potential tech debt) vs coming up with novel form patterns.

## Dependencies



## Screenshots

<img width="1212" alt="screen shot 2019-02-14 at 5 16 51 pm" src="https://user-images.githubusercontent.com/19582796/52828129-70ea6c00-307c-11e9-89fa-85a2d876b317.png">
<img width="1210" alt="screen shot 2019-02-14 at 5 17 02 pm" src="https://user-images.githubusercontent.com/19582796/52828139-78117a00-307c-11e9-847c-4d597cde6628.png">
<img width="1210" alt="screen shot 2019-02-14 at 5 17 18 pm" src="https://user-images.githubusercontent.com/19582796/52828153-7fd11e80-307c-11e9-96ae-f8903ef27783.png">
